feat(FileUploader): add drag and drop file uploader

emyarod opened the pull request:
Closes #2088 
Closes #2916
Closes #2917 
Closes #3252

refs #2848 #3016 #3175

This PR adds drag and drop support to the React file uploader component.

#### Changelog

**New**

- `<FileUploaderDropContainer>` component to represent drop area for file uploads
- `<FileUploaderItem>` component to represented uploaded items
- example application showing drag and drop file uploader workflow, error handling, etc

**Changed**

- correct file upload status icons
- match invalid file upload markup structure with loading and success upload structures
- deprecate legacy styles
- convert `<Filename>` to stateless functional component

**Removed**

- replace hard coded loader with `<InlineLoading />` (to make use of #2955)

#### Testing / Reviewing

Ensure the new file uploader functions as expected and ensure the existing file loader does not have any regressions

asudoh reviewed (approved):
LGTM 👍 - Thanks @emyarod!

aledavila reviewed (changes requested):
1) Disabled state functionally works however there is no actual visible disabled state to the component so I wouldn't know as a user its actually disabled rather than broken. Is there a variation for disabled state ? 

2) Do we need to have IBM copyright at the top of every file ? We don't have that in any of these. 

3) Are all the props not represented in prop types on purpose ?

emyarod commented:
1. I added disabled styles but since we don't have a spec for it cc @shixiedesign 

2. added copyright headers

3. which props are you referring to? for instance, not all rest/spread props will have defined types

aledavila commented:
@emyarod It was more of curiosity for example disabled isn't defined in PropTypes. Thanks for the other updates. I'm good once design clarifies disabled state of it.

shixiedesign reviewed (commented):
Hi @emyarod looks like we are re-doing the reviews so I took a look at the component in general and noticed a few things: 

- [x] Hover state, the link inside should gain the underline, but stay blue, using our `$link-01` color.

![Aug-30-2019 12-07-48](https://user-images.githubusercontent.com/15144993/64038864-e4d2fa00-cb1e-11e9-93e7-1def21cbbc04.gif)

- [x] Active state, the link inside should turn black, using `$text-01` color, basically getting the current hover styling

- [x] Disabled state, sorry for missing the spec on that! See below for active & disabled states spec: 

![image](https://user-images.githubusercontent.com/15144993/64043183-4dbf6f80-cb29-11e9-97cc-5af47afc51e7.png)

emyarod commented:
should be resolved @shixiedesign

vpicone commented:
@emyarod I'm by what's confused by the story. It seems random whether or not the file gets marked as failed or not? Uploading the same 116kb jpeg twice in a row and one of them got rejected while the other didn’t.


<img width="373" alt="Screen Shot 2019-08-30 at 6 35 37 PM" src="https://user-images.githubusercontent.com/4078018/64055899-2d0d0f00-cb55-11e9-81f4-227541b76dd6.png">

vpicone commented on packages/react/src/components/FileUploader/FileUploaderDropContainer.js:
```diff
@@ -0,0 +1,173 @@
+            ref={inputRef}
+            tabIndex="-1"
+            disabled={disabled}
+            accept={accept}
```
This only works for if someone clicks to upload. Shouldn't we be validating the extensions for files added through drag and drop as well?

emyarod replied:
added file extension and mime type checks for drop uploads

emyarod commented:
@vpicone the error messages in the story are configurable via knobs and aren't meant to reflect the random failures in our mock uploads. in any case I modified the default knobs

vpicone commented:
@emyarod so the failures are random? That seems like it would be confusing to someone testing error message behavior/appearance

emyarod commented:
@vpicone in the demo yes. do you propose that we just validate using a specific rule?

vpicone commented:
@emyarod yes, the randomness confused me I imagine it would confuse someone without the story source open even more.

We should actually implement the validation behavior not just make it fail randomly

vpicone reviewed (changes requested):
@emyarod How are they meant to provide the error message subject/body without the size/file extension? Should those be render functions rather than text nodes? 

Are the only allowed to have one kind of error?

emyarod commented:
@vpicone what should the example error be? right now I have it as just a server error

the File objects are sent to the application so all of the file information can be found in the example app state

shixiedesign reviewed (commented):
All earlier comments were addressed! Need two last fixes: 

- [x] animation of the hover seems have a glitch
Current:
![Sep-03-2019 10-20-02](https://user-images.githubusercontent.com/15144993/64193532-879eb780-ce42-11e9-8dc3-d594bc3fd12f.gif)
Should be:
![Sep-03-2019 11-17-30](https://user-images.githubusercontent.com/15144993/64193536-8c636b80-ce42-11e9-9354-d9ad51a8da5a.gif)

- [x] Content inside drop zone ("Drag and drop files here or click to upload") should be top-left aligned, 16px from corner, like so: 
![image](https://user-images.githubusercontent.com/15144993/64193607-ba48b000-ce42-11e9-97ed-9885dcf524b8.png)

emyarod commented:
@vpicone rather than a separate prop that takes a validator function, I added the validation logic to the example upload method

shixiedesign reviewed (approved):
This looks good! Thank you Andrew!

vpicone commented:
Should we be using text overflow here? Should the X be stuck to the top?
<img width="419" alt="Screen Shot 2019-09-04 at 11 24 52 AM" src="https://user-images.githubusercontent.com/4078018/64273465-fcd3c080-cf06-11e9-953d-57ad4cc03792.png">

vpicone reviewed (changes requested):
The story source does not properly document the new component, we need prop types and information on using the component itself. The example here, if we keep it, needs to have proper language rather than 'lorem ipsum'
 
https://deploy-preview-3872--carbon-components-react.netlify.com/?path=/story/fileuploader--drag-and-drop-upload-container

shixiedesign commented:
Great catch @vpicone – for long file names, we are hoping to achieve this kind of truncation, where we truncate the name but keep the extension. However @asudoh worked on this a while back and got back to me saying keeping the truncation is hard, so we were hoping to just achieve the simple truncation at the moment: 

- [x] Fix file list item height, truncate file names as below:
![image](https://user-images.githubusercontent.com/15144993/64275231-c009c880-cf0a-11e9-9d16-67b166907ad9.png)

Another couple of style issue I noticed, after seeing the screenshot, @emyarod

- [x] the error messages (both red error message and the additional message in gray) should be 12px
- [x] File name text should be 14px, using `$body-short-01` 
- [x] Divider line should span the whole width of the file list item
- [x] Error state is missing the Warning icon in `$support-01`

**Current:** 
![image](https://user-images.githubusercontent.com/15144993/64275645-b7fe5880-cf0b-11e9-9ea4-94f11ae126f0.png)
**Should be:**
![image](https://user-images.githubusercontent.com/15144993/64275661-c187c080-cf0b-11e9-88cf-161c395e6276.png)

vpicone commented:
Nice, that’s better can we the story name to be the component so it matches FileUploader and FileUploader button, and sets it apart from the example story.

Also we include a story for `FileUploaderItem` with and without errors (even if it’s hard coded) so those props are documented and someone can readily see an error and normal item. They can be totally hard coded, just purely presentational representations of those components

emyarod commented:
edited the dnd file uploader story and added a story for `FileUploaderItem`. planning to migrate the example application to a clearer codesandbox example once the component is merged

vpicone commented on packages/components/src/components/file-uploader/_file-uploader.scss:
```diff
@@ -54,85 +101,164 @@
     }
+
+    .#{$prefix}--file-filename {
+      margin-left: rem(16px);
```
```suggestion
      margin-left: $carbon--spacing-05;
```

vpicone commented on packages/components/src/components/file-uploader/_file-uploader.scss:
```diff
@@ -54,85 +101,164 @@
+
+  .#{$prefix}--file__selected-file--invalid .#{$prefix}--form-requirement {
+    border-top: 1px solid $ui-03;
+    padding-top: rem(8px);
```
spacing token

vpicone commented on packages/components/src/components/file-uploader/_file-uploader.scss:
```diff
@@ -54,85 +101,164 @@
+
+  .#{$prefix}--file__selected-file--invalid
+    .#{$prefix}--form-requirement__title {
+    padding-right: rem(16px);
```
spacing tokens for all these

vpicone commented on packages/components/src/components/file-uploader/_file-uploader.scss:
```diff
@@ -54,85 +101,164 @@
+  }
+
+  .#{$prefix}--file__state-container .#{$prefix}--file-invalid {
+    height: rem(16px);
```
here too

vpicone reviewed (changes requested):
We need to replace all of the rem(16/8/12px) with the appropriate spacing tokens

vpicone commented on packages/components/src/components/file-uploader/_file-uploader.scss:
```diff
@@ -54,85 +101,164 @@
+    // for checkmark contrast
+    [data-icon-path='inner-path'] {
+      opacity: 1;
+      fill: $ibm-color__white-0;
```
deprecated token, also should this match the background color/be a theme color instead of plain white for themabality?

emyarod replied:
not sure what token should be used but this is based on feedback from @shixiedesign https://github.com/carbon-design-system/carbon/pull/3175#pullrequestreview-254372192

vpicone replied:
At the very least we need to be using the carbon one not the deprecated IBM prefix.

shixiedesign replied:
The checkmark should be white through all themes, let's use `$icon-03`!

vpicone commented on packages/components/src/components/file-uploader/_file-uploader.scss:
```diff
@@ -54,85 +101,165 @@
+    .#{$prefix}--form-requirement__supplement {
+    padding-right: $carbon--spacing-05;
+    padding-left: $carbon--spacing-05;
+    font-size: rem(12px);
```
@shixiedesign shouldn't we be using the `label-01` type token rather than just setting the font size to 12px?

> This is for field labels in components and error messages.

https://www.carbondesignsystem.com/guidelines/typography/productive

shixiedesign replied:
Yep that's right. Damn Vince you are on fire! This is my fault I realize I didn't spec any of the type! @emyarod  Here are the type tokens in this component: 

![image](https://user-images.githubusercontent.com/15144993/64286655-90b38580-cf23-11e9-9702-71425d1b4b91.png)

shixiedesign commented:
Just noticed that the loading spinner is slightly shifted to the left than the checkmark / x icon. They should be aligned: 
**Current:**
![Sep-04-2019 14-54-48](https://user-images.githubusercontent.com/15144993/64286910-fef84800-cf23-11e9-9856-856d96294ad5.gif)
**Should be**
![image](https://user-images.githubusercontent.com/15144993/64287064-5696b380-cf24-11e9-97e6-ebc8dce052ce.png)

vpicone commented on packages/components/src/components/file-uploader/_file-uploader.scss:
```diff
@@ -54,85 +101,165 @@
+    justify-content: space-between;
+    height: rem(96px);
+    max-width: rem(320px);
+    padding: 16px;
```
```suggestion
    padding: $carbon--spacing-05;
```

emyarod committed:
feat(FileUploader): add drag and drop file uploader

vpicone reviewed (approved):
Great work on this

joshblack committed:
Merge branch 'master' into dnd-file-uploader-0

joshblack committed:
Merge branch 'master' into dnd-file-uploader-0
